Sources/HttpPipeline/Middleware.swift

```diff
@@ -1,61 +1,68 @@
 import Foundation
+import Optics
 import Prelude
 
 public typealias Middleware<I, J, A, B> = (Conn<I, A>) -> Conn<J, B>
 
 public func writeStatus<A>(_ status: Status) -> Middleware<StatusLineOpen, HeadersOpen, A, A> {
-  return { connection in
-    return .init(
-      data: connection.data,
-      request: connection.request,
-      response: Response(status: status, headers: connection.response.headers, body: nil)
-    )
-  }
-}
-
-public func writeHeader<A>(_ header: ResponseHeader) -> Middleware<HeadersOpen, HeadersOpen, A, A> {
   return { conn in
-    .init(
+    return .init(
       data: conn.data,
       request: conn.request,
       response: Response(
-        status: conn.response.status,
-        headers: conn.response.headers + [header], body: nil
+        status: status,
+        headers: conn.response.headers,
+        body: nil
       )
     )
   }
 }
 
+public func writeHeader<A>(_ header: ResponseHeader) -> Middleware<HeadersOpen, HeadersOpen, A, A> {
+  return \.response.headers %~ { $0 + [header] }
+}
+
 public func writeHeader<A>(_ name: String, _ value: String) -> Middleware<HeadersOpen, HeadersOpen, A, A> {
   return writeHeader(.other(name, value))
 }
 
-public func closeHeaders<A>() -> Middleware<HeadersOpen, BodyOpen, A, A> {
-  return { conn in
-    .init(
-      data: conn.data,
-      request: conn.request,
-      response: conn.response
-    )
-  }
+public func closeHeaders<A>(conn: Conn<HeadersOpen, A>) -> Conn<BodyOpen, A> {
+  return .init(
+    data: conn.data,
+    request: conn.request,
+    response: conn.response
+  )
 }
 
-public let end: Middleware<BodyOpen, ResponseEnded, Data?, Data?> =
-  { conn in
-    .init(
-      data: conn.data,
-      request: conn.request,
-      response: Response(
-        status: conn.response.status,
-        headers: conn.response.headers,
-        body: conn.data
-      )
+public func end(conn: Conn<BodyOpen, Data?>) -> Conn<ResponseEnded, Data?> {
```

Comment:
made `end` into a function so that there could be two overloads. this is the normal one...